src/examples/libpmemobj/queue/queue.c

```diff
@@ -0,0 +1,279 @@
+{
+	return POBJ_ALLOC(pop, q, struct queue,
+		sizeof(struct queue) + sizeof(TOID(struct entry)) * nentries,
+		queue_constructor, &nentries);
+}
+
+/*
+ * queue_nentries -- returns the number of entries
+ */
+static size_t
+queue_nentries(struct queue *queue)
+{
+	return queue->rear - queue->front;
+}
+
+/*
+ * queue_push -- allocates and inserts a new entry into the queue
+ */
+static int
+queue_push(PMEMobjpool *pop, struct queue *queue, char *data, size_t len)
+{
+	if (queue->capacity - queue_nentries(queue) == 0)
+		return -1; /* at capacity */
+
+	/* rear is never decreased, need to calculate the real position */
+	size_t pos = queue->rear & (queue->capacity - 1);
+
+	int ret = 0;
+
+	TX_BEGIN(pop) {
+		/* let's first reserve the space at the end of the queue */
+		TX_ADD_FIELD_DIRECT(queue, rear);
+		queue->rear += 1;
+
+		/* and then snapshot the queue entry that we will allocate to */
+		pmemobj_tx_add_range_direct(&queue->entries[pos],
+			sizeof(TOID(struct entry)));
+
+		/* now we can safely allocate and initialize the new entry */
+		queue->entries[pos] = TX_ALLOC(struct entry,
+			sizeof(struct entry) + len);
+		memcpy(D_RW(queue->entries[pos])->data, data, len);
+	} TX_ONABORT { /* don't forget about error handling! ;) */
+		ret = -1;
+	} TX_END
+
+	return ret;
+}
+
+/*
+ * queue_pop - removes and frees the first element from the queue
+ */
+static int
+queue_pop(PMEMobjpool *pop, struct queue *queue)
+{
+	if (queue_nentries(queue) == 0)
+		return -1; /* no entries to remove */
+
+	/* front is also never decreased */
+	size_t pos = queue->front & (queue->capacity - 1);
+
+	int ret = 0;
+
+	TX_BEGIN(pop) {
+		/* move the queue forward */
+		TX_ADD_FIELD_DIRECT(queue, front);
+		queue->front += 1;
+		/* and since this entry is now unreachable, free it */
+		TX_FREE(queue->entries[pos]);
+	} TX_ONABORT {
+		ret = -1;
+	} TX_END
+
+	return ret;
+}
+
+/*
+ * queue_show -- prints all queue entries
+ */
+static void
+queue_show(PMEMobjpool *pop, struct queue *queue)
+{
+	size_t nentries = queue_nentries(queue);
+	printf("Entries %zu/%zu\n", nentries, queue->capacity);
+	for (size_t i = queue->front; i < queue->rear; ++i) {
+		size_t pos = i & (queue->capacity - 1);
+		printf("%zu: %s\n", pos, D_RW(queue->entries[pos])->data);
+	}
+}
+
+/* available queue operations */
+enum queue_op {
+	UNKNOWN_QUEUE_OP,
+	QUEUE_NEW,
+	QUEUE_PUSH,
+	QUEUE_POP,
+	QUEUE_SHOW,
+
+	MAX_QUEUE_OP,
+};
+
+/* queue operations strings */
+const char *ops_str[MAX_QUEUE_OP] = {"", "new", "push", "pop", "show"};
+
+/*
+ * parse_queue_op -- parses the operation string and returns matching queue_op
+ */
+static enum queue_op
+queue_op_parse(const char *str)
+{
+	for (int i = 0; i < MAX_QUEUE_OP; ++i)
+		if (strcmp(str, ops_str[i]) == 0)
+			return (enum queue_op)i;
+
+	return UNKNOWN_QUEUE_OP;
+}
+
+/*
+ * fail -- helper function to exit the application in the event of an error
+ */
+static void __attribute__((noreturn))
```

Comment:
__attribute__((noreturn)) is not needed (all static analyzers will notice that the last function is marked with this attribute) and may be confusing to newbies.

Reply:
while it is not required for gcc, clang complains about it.